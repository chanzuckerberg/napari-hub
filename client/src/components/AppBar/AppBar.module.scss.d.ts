export type Styles = {
  appBar: string;
  header: string;
  links: string;
  menuButton: string;
  searchContainer: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
