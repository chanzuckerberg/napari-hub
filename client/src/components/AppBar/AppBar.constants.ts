import { AppLink } from '@/types';

export const APP_LINKS: AppLink[] = [
  {
    title: 'About',
    link: '/about',
  },
  {
    title: 'FAQ',
    link: '/faq',
  },
];
