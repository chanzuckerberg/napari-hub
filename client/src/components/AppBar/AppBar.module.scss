.links {
  :not(:first-child) {
    @apply ml-6;
  }
}
