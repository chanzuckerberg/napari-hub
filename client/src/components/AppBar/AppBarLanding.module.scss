.heading {
  font-size: clamp(35px, 10vw, 61px);
}

.logo {
  width: clamp(122px, 35vw, 223px);
}

.list li {
  font-size: clamp(14px, 4vw, 24px);
}
