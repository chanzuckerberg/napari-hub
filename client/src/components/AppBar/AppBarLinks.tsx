import clsx from 'clsx';

import { Link } from '@/components/common';
import { useSearchState } from '@/context/search';
import { AppLink } from '@/types';

interface CommonProps {
  vertical?: boolean;
}

interface LinkListProps extends CommonProps {
  items: AppLink[];
}

/**
 * Link bar for rendering menu links. This only shows up on lg+ screens.
 */
function LinkList({ items, vertical }: LinkListProps) {
  return (
    <div className={clsx('flex', vertical && 'flex-col')}>
      {items.map((item) => (
        <Link
          className={clsx(vertical ? 'mt-4' : 'ml-6')}
          key={item.link}
          href={item.link}
        >
          {item.title}
        </Link>
      ))}
    </div>
  );
}

interface Props extends CommonProps {
  className?: string;
  items?: AppLink[];
}

/**
 * Component that includes site-wide navigational links.
 */
export function AppBarLinks({ className, items, vertical }: Props) {
  const { setQuery } = useSearchState() ?? {};

  return (
    <nav className={clsx(className, 'flex', vertical && 'flex-col')}>
      <h1 className="whitespace-nowrap">
        <Link
          // Redirect to home page
          href="/"
          // Clear search related query parameter data if the user is currently
          // on the search page. Without this, the `useQueryParameter()` hook
          // will re-set the query parameter with the current query in the
          // search bar.
          onClick={() => setQuery?.('')}
        >
          napari <strong>hub</strong>
        </Link>
      </h1>

      {items && <LinkList items={items} vertical={vertical} />}
    </nav>
  );
}
