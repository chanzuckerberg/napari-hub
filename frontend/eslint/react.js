module.exports = {
  extends: ['airbnb/hooks'],

  plugins: ['napari-hub'],

  rules: {
    'napari-hub/no-root-mui-import': 'error',

    /*
      Prop types aren't necessary since we have TypeScript interfaces for
      prop types.
    */
    'react/prop-types': 'off',

    /*
      React automatically adds the import as part of the new JSX transform:
      https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#nextjs
     */
    'react/react-in-jsx-scope': 'off',

    /*
      Prop spreading is very useful as a simple way for passing multiple
      props to a component. However, we should still be careful about passing
      unnecessary props.
    */
    'react/jsx-props-no-spreading': 'off',

    /*
      This rule isn't really needed anymore since we can use ES6 object defaults
      for props. There are also plans to deprecate the use of `defaultProps`:
      https://github.com/reactjs/rfcs/pull/107
    */
    'react/require-default-props': 'off',
  },
};
