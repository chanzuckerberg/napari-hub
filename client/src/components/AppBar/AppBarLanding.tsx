import clsx from 'clsx';
import { ReactNode } from 'react-markdown';

import { ColumnLayout } from '@/components/common';
import { Hub } from '@/components/common/icons';

import { Media, MediaFragment } from '../common/media';
import { APP_LINKS } from './AppBar.constants';
import styles from './AppBarLanding.module.scss';
import { AppBarLinks } from './AppBarLinks';

interface ListProps {
  children: ReactNode;
  className?: string;
}

function List({ children, className }: ListProps) {
  return (
    <ul
      className={clsx(
        'list-disc font-semibold space-y-4',
        styles.list,
        className,
      )}
    >
      {children}
    </ul>
  );
}

/**
 * Component that renders the landing page variant of the AppBar.
 */
export function AppBarLanding() {
  // On smaller layouts, the last list item is rendered below the icon and 2
  // preceding list items. To achieve this affect, the last list item is
  // rendered in a separate list so that it can be rendered below everything.
  const lastListNode = (
    <li>Share your image analysis tools with napariâ€™s growing community</li>
  );

  return (
    <ColumnLayout
      className="bg-napari-primary p-6 screen-495:p-12"
      classes={{
        // Use 3-column layout instead of 4-column.
        fourColumn: 'screen-1150:grid-cols-napari-3',
      }}
      component="header"
    >
      <Media greaterThanOrEqual="screen-1425">
        <AppBarLinks items={APP_LINKS} vertical />
      </Media>

      <Media lessThan="screen-1425">
        <AppBarLinks className="mb-6" items={APP_LINKS} />
      </Media>

      <h1
        className={clsx(
          'font-bold col-span-full',
          '-mt-12 screen-1425:m-0',
          'screen-1425:col-start-2 screen-1425:col-span-3',
          styles.heading,
        )}
      >
        Discover, install, and share napari plugins
      </h1>

      <Hub
        className={clsx(
          'h-full',
          'self-center',
          'screen-1425:row-start-2 screen-1425:col-start-2',
          styles.logo,
        )}
      />

      <List className="ml-6 screen-875:col-span-2">
        <li>Discover plugins that solve your image analysis challenges</li>
        <li>Install into napari straight from your browser</li>

        {/* Render in list if screen is wide enough. */}
        <MediaFragment greaterThanOrEqual="screen-725">
          {lastListNode}
        </MediaFragment>
      </List>

      {/* Render as separate list so that it renders below everything on smaller screens. */}
      <MediaFragment lessThan="screen-725">
        <List className="col-span-2 px-6">{lastListNode}</List>
      </MediaFragment>
    </ColumnLayout>
  );
}
