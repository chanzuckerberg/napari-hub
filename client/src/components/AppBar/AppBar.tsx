import clsx from 'clsx';
import Image from 'next/image';
import { useState } from 'react';

import { MenuDrawer, SearchBar } from '@/components';
import { Link } from '@/components/common';
import { MenuDrawerItem } from '@/components/MenuDrawer/types';
import { useSearchState } from '@/context/search';

import styles from './AppBar.module.scss';

const IMAGE_SIZE = 16;

const MENU_ITEMS: MenuDrawerItem[] = [
  {
    title: 'About',
    link: '/about',
  },
  {
    title: 'Help',
    link: '/help',
  },
];

/**
 * Header that links back to the home page.
 */
function AppBarHeader() {
  const { setQuery } = useSearchState() ?? {};

  return (
    <header className="flex">
      <h1 className="whitespace-nowrap">
        <Link
          // Redirect to home page
          href="/"
          // Clear search related query parameter data if the user is currently
          // on the search page. Without this, the `useQueryParameter()` hook
          // will re-set the query parameter with the current query in the
          // search bar.
          onClick={() => setQuery?.('')}
        >
          napari <strong>hub</strong>
        </Link>
      </h1>
    </header>
  );
}

/**
 * Link bar for rendering menu links. This only shows up on lg+ screens.
 */
function AppBarLinks() {
  return (
    <ul
      className={clsx(
        // Hide links on smaller layouts
        'hidden lg:flex',

        // Margins
        'ml-12',

        // Custom link styling
        styles.links,
      )}
    >
      {MENU_ITEMS.map((item) => (
        <li className="list-none" key={item.link}>
          <Link href={item.link}>{item.title}</Link>
        </li>
      ))}
    </ul>
  );
}

/**
 * App bar component that renders the home link, search bar, and menu.
 */
export function AppBar() {
  const [visible, setVisible] = useState(false);

  return (
    <>
      <MenuDrawer
        items={MENU_ITEMS}
        onMenuClose={() => setVisible(false)}
        visible={visible}
      />

      <nav
        className={clsx(
          // Color and height
          'bg-napari-primary h-napari-app-bar',

          // Padding
          'px-6 md:px-12 2xl:p-0',

          // Grid layout
          'grid grid-cols-napari-nav-mobile',
          'justify-center items-center',

          // Grid gap
          'gap-6 md:gap-12',

          // Change to 2 column grid layout when 2xl+ screens
          '2xl:grid 2xl:grid-cols-napari-app-bar-2-col',

          // Use 3 column layout when 3xl+ screens
          '3xl:grid 3xl:grid-cols-napari-3-col',
        )}
      >
        <AppBarHeader />

        <div
          className={clsx(
            // Flex layout
            'flex items-center',

            // Take 100% of width, but limit to center column width.
            'w-full max-w-napari-center-col',

            // Align container to the right of the grid cell
            'justify-self-end',
          )}
        >
          <SearchBar />
          <AppBarLinks />

          {/* Menu button */}
          <button
            // Show menu button on smaller layouts
            className="ml-6 flex lg:hidden"
            onClick={() => setVisible(true)}
            type="button"
          >
            <Image
              src="/icons/menu.svg"
              alt="Menu button icon"
              width={IMAGE_SIZE}
              height={IMAGE_SIZE}
            />
          </button>
        </div>
      </nav>
    </>
  );
}
