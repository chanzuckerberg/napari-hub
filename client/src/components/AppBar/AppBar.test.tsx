import { render } from '@testing-library/react';

import { AppBar } from './AppBar';

describe('<AppBar />', () => {
  it('should match snapshot', () => {
    const component = render(<AppBar />);
    expect(component.asFragment()).toMatchSnapshot();
  });
});
