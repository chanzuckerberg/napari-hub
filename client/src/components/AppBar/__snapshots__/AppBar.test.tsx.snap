// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppBar /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="fixed top-0 right-0 flex flex-row items-start z-40 w-9/12 h-screen p-6 bg-black"
    data-testid="menu"
    style="transform: translateX(100%) translateZ(0);"
  >
    <ul
      class="flex flex-auto flex-col"
    >
      <li
        class="text-white item"
        data-testid="drawerItem"
      >
        <a
          href="/about"
        >
          About
        </a>
      </li>
      <li
        class="text-white item"
        data-testid="drawerItem"
      >
        <a
          href="/faq"
        >
          FAQ
        </a>
      </li>
    </ul>
    <button
      class="flex"
      data-testid="drawerClose"
      type="button"
    >
      <svg
        fill="none"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1 1L17 17"
          stroke="white"
          stroke-width="2.35294"
        />
        <path
          d="M17 1L1 17"
          stroke="white"
          stroke-width="2.35294"
        />
      </svg>
    </button>
  </div>
  <header
    class="bg-napari-primary h-napari-app-bar justify-center items-center px-6 md:px-12 2xl:p-0 zero:grid-cols-[min-content,1fr] grid justify-center gap-6 md:gap-12 grid-cols-2 screen-875:grid-cols-napari-3 screen-1150:grid-cols-napari-4 screen-1425:grid-cols-napari-5"
  >
    <div
      class="fresnel-container fresnel-greaterThanOrEqual-lg "
    >
      <nav
        class="flex"
      >
        <h1
          class="whitespace-nowrap"
        >
          <a
            href="/"
          >
            napari 
            <strong>
              hub
            </strong>
          </a>
        </h1>
        <div
          class="flex"
        >
          <a
            class="ml-6"
            href="/about"
          >
            About
          </a>
          <a
            class="ml-6"
            href="/faq"
          >
            FAQ
          </a>
        </div>
      </nav>
    </div>
    <div
      class="fresnel-container fresnel-lessThan-lg "
    >
      <nav
        class="flex"
      >
        <h1
          class="whitespace-nowrap"
        >
          <a
            href="/"
          >
            napari 
            <strong>
              hub
            </strong>
          </a>
        </h1>
      </nav>
    </div>
    <div
      class="flex items-center w-full justify-self-end xl:col-span-2 2xl:col-span-3"
    >
      <form
        class="flex flex-auto items-center border-b-2 border-black"
        data-testid="searchBarForm"
      >
        <input
          class="flex flex-auto border-none outline-none bg-transparent w-0"
          data-testid="searchBarInput"
          value=""
        />
        <svg
          class="w-5 h-5"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.32004 7.68001L0.746704 15.2533M10.6667 9.60001C13.0134 9.60001 14.9334 7.68001 14.9334 5.33335C14.9334 2.98668 13.0134 1.06668 10.6667 1.06668C8.32004 1.06668 6.40004 2.98668 6.40004 5.33335C6.40004 7.68001 8.32004 9.60001 10.6667 9.60001Z"
            stroke="black"
            stroke-width="2"
          />
        </svg>
      </form>
      <button
        class="ml-6 flex lg:hidden"
        type="button"
      >
        <svg
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 1H0"
            stroke="black"
            stroke-width="1.98857"
          />
          <path
            d="M16 7.99924H0"
            stroke="black"
            stroke-width="1.98857"
          />
          <path
            d="M16 15H0"
            stroke="black"
            stroke-width="1.98857"
          />
        </svg>
      </button>
    </div>
  </header>
</DocumentFragment>
`;
